policies/activation_checkpointing_functions.py

```diff
@@ -31,3 +32,65 @@ def apply_fsdp_checkpointing(model):
     apply_activation_checkpointing(
         model, checkpoint_wrapper_fn=non_reentrant_wrapper, check_fn=check_fn
     )
+
+
+class FreeEventQueue:
+    """https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_limiter_utils.py"""
+    def __init__(self):
+        self._queue = collections.deque()
+        self._max_num_inflight_copy = 2
+
+    def enqueue(self, free_event):
+        self._queue.append(free_event)
+
+    def dequeue_if_needed(self):
+        if len(self._queue) >= self._max_num_inflight_copy:
+            return self._dequeue()
+        return None
+
+    def _dequeue(self):
+        if self._queue:
+            event = self._queue.popleft()
+            return event
+        return None
+
+
+class save_on_cpu_overlap(torch.autograd.graph.saved_tensors_hooks):
```

Comment:
I am not sure how this works and if it is safe tbh.
- What are you using events for here? Why is the very first thing you do dequeue from an empty queue?
- How do you ensure the CPU waits for the copy to be ready in packed? Can you guarantee no one will touch that Tensor until then?
-

Reply:
@albanD Thanks for your comments.

> What are you using events for here? Why is the very first thing you do dequeue from an empty queue?

The rationale is the same with that used in FSDP, described in the paper [PyTorch FSDP](https://arxiv.org/abs/2304.11277), see Section 3.4:

"FSDP offers a rate limiter that intentionally blocks the CPU thread to ensure proper caching allocator block reuse. It allows at most two inflight AllGathers, which is the minimum amount to still achieve communication and computation overlap."

Similarly, we maintain a sliding window here for the inflight copy operations for the sake of memory.

> How do you ensure the CPU waits for the copy to be ready in packed? Can you guarantee no one will touch that Tensor until then?

IIUC, the packed tensor on CPU will not be touched by users, and will be unpacked in the registered hook with proper synchronization :)

Reply:
> Similarly, we maintain a sliding window here for the inflight copy operations for the sake of memory.

That sounds like a lot of CPU syncs. That might be pretty bad once you have real workloads as it will prevent the CPU from being ahead.

>  the packed tensor on CPU will not be touched by users

Nothing guarantees that?

> be unpacked in the registered hook with proper synchronization :)

The stream use and wait seems to be good idd!
Note that the record_stream calls can lead to non-deterministic behavior but that fine in terms of correctness.